### ISSUE TITLE
Passing graphs like functions are passed in functional programming 

### ISSUE THREAD

**Issue by seefeldb** [google/labs-prototypes#110]
We currently passing graphs in `include` and `slot`, and that combination if particular powerful: It's almost like passing a function to a function. For example you could imagine a mapOverArray.graph that has a slot for the operation that should happen over each element.

But right now both graphs are self enclosed and not connected to the calling graph. What if we could wire things more or less directly from the calling graph into these passed graphs. It might be a terrible idea, but let's have a look.

Something like `query = "..."; filtered = list.filter(x => x === query)`, where the passed function can access `query` from the closure could become possible.

Two approaches:

## 1) Where we add functions

We could add a way to define a subgraph, but then also wire `query` directly from the context into this subgraph, then pass that subgraph. This could look like this:

```
query = board.input();
node.wire("list->",
  kit.filter()
    .wire("slot<-",
      board.input()
        .wire("x->", kit.compare()
          .wire("y<-.query", query)
          .wire("x->", board.output())))
    .wire()....;
```

(I invented a `kit.compare` node to save typing, that's not part of the suggestion)

It's pretty close to how we'd write it today, except for not serializing to JSON before sending it.

This has implications on #63 (dynamic includes). Also note, re #108, if the wire from query wasn't a constant ('<-.`), this would only work for one round. OTOH, this is how one could write a way to merge streams, though it seems a little shaky...

Note that when serializing this, it's all one graph, with the `filter` node getting a reference to the new `input` node and the `compare` node already wired up with a queued up opportunity. This is a key difference to how `include` and `slot` are implemented: We'd have to reuse the same `GraphRunner` instance, or otherwise have shared queues of opportunities.

Maybe that's how `include` should be implemented even when loading graphs, with the main difference between graphs being their own namespaces for ids and such. This might also eventually help with running things in parallel (though maybe not necessary, we'd just have multiple graph runners running in parallel, with less control on scheduling).

## 2) Using * to pass remainder inputs to subgraphs

This requires less changes. Feels a bit hacky, but maybe it's not: `filter` (whether as actual node or implemented as graph!) just passes all extra incoming wires to its slotted graphs, e.g. here `query`.

```
const innerBoard = new Board();
const compare = kit.compare();

// Side node: I wish this was easier, maybe .wire("[x,query]->" or .wire(["x->", "query->"]?
innerBoard.input().
  .wire("x->", compare)
  .wire("query->", compare)
compare.wire("x->", board.output())

query = board.input();
node.wire("list->",
  kit.filter()
    .wire("query<-", query)
    .wire("slot<-", innerBoard as unknown as NodeValue)
    .wire()....;
```

This makes it impossible for the inner board to output anything outside of `output`, where as above we could add wires out at any point (unlike functional programming!).

All this needs is a way to capture the "remainder" inputs as `*` somehow. Maybe that's not even necessary, it'll just keep passing `list` and `slot` to the inner board, where it gets ignored. But this makes the issue with this approach more apparent: There is potential for name collisions, especially if the operators (map, filter, etc) are being developed independently of what can go inside.

### 2b) bundler nodes?

To address that last issue and create a namespace we could create bundler nodes that either collect `*` into a single value or expand such a single value into a `*`. 

So
```
...
   .wire("slotArgs<-.bundle", board.bundle().wire("<-query", query))
```
in the caller (with a convention that matches up slot names with arguments)

And
```
...
   .wire("<-query", board.unbundle().wire("bundle<-args").input)
...
```
in the function (with a convention that this is passed as `args`).

A bit more typing, but could work.

Alternatively, skip that second change in the function and do pass all args expanded already. The namespace issue is really more in the operator and the bundle tunnels through that. Then it's just a few extra characters to bundle the inputs.

## 3) `bind` as a mix of the two, and probably the best option

Here we bind a set of inputs to a graph representation (or board) and then pass that as slot input. Once the runner gets to this board, it adds these inputs to the ones passed in by the caller with the caller inputs getting precedence.

Syntax TBD.

FWIW:

A nicer syntax for either might be to have kit.map() accept a (JavaScript) function that gets board.input() and board.output() as parameters, with the slot made implicit:

```
query = board.input();
node.wire("list->",
  kit.filter((in, out) => in.wire("x->", kit.compare()
          .wire("y<-.query", query) // or `in` for the second proposal
          .wire("x->", out))
    .wire()....;
```

or something like that to make it all feel more like regular javascript code.

**Comment by seefeldb** [comment]
One concrete example could be generalizing schemish-generator, @dglazkov :

That graph could have a slot for the actual generation, including prompt generation (with space for feedback from the checker for future rounds). Then the logic is just about checking the schema, and we can plug in any kind of generator there.

Similarly for other generator-critic patterns: the critic wraps the generator. It could even be multi level: A plan critic that requires valid JSON wraps a schemish graph that includes the actual generator via slot. Functional programming!

In all of these, we need to wire data to the slotted graphs. So we'll need something like above.

**Comment by dglazkov** [comment]
One interesting quirk here is that graphs are asynchronous generators. This seems like a really interesting uplevel of the whole functional programming bit.

**Comment by seefeldb** [comment]
Yes! The simple scenario is where the bound inputs (in variant 3) are like constant inputs and the outputs are collected along the chain. But we can do so many other things:
 - consume inputs, so work down a queue. A little brittle though, especially if we have to retry a few steps or so.
 - write out into other wires without the caller being involved. That's quite interesting, as those can trigger other flows, etc.

This makes we wonder though whether critics should indeed be wrappers as described above. Maybe a decorator pattern is better, syntactically? The problem though is that what they decorate is a distinct graph. Either way, what we'd want is that the generator (in the LLM sense) is a generator (in the PL sense), that the decoration is seamless at the right level.

Let's play with it. In the meantime I think the `bind` (or `curry`?) approach is worth doing. WDYT?

**Comment by dglazkov** [comment]
Also the generator bit seems to be related to queueing and streaming.

**Comment by dglazkov** [comment]
What if nodes are also async generators?

**Comment by dglazkov** [comment]
I am not grokking the "bind" approach. I tried reading, but it's still not clicking.

**Comment by seefeldb** [comment]
I just realized that the syntactically, the only thing we need to add is to
support the idea is to allow wires to boards that are passed in an include
as slots:

If you split schemish-generator in two parts, make a graph that assembles
the prompt and wire prologue and so on directly to the board. Pass that
board and the schema to the schemish-generator board. Done.

The schemish- part can be connected to any other generator in the same way.

And we should be able to do a call-generated-function that feeds error
messages back to a code generation graph.


For `include`:

If the graph is included exactly once, then there is no difference between
wiring to the board and including it or wiring inputs to the include node.
In fact you'll still need something to go to include node to trigger the
run.

But if you include the same graph multiple times, then this is a way to
factor out wires that need to be present each time! E.g. a secret or so.

So in the sprit of graph and node equivalence, maybe we should support that
for nodes as well? Bind secrets to generateText and get a new node type
where you don't have to wire the secret every single time.

On the phone right now, but I can type out the examples as code if that
helps.

**Comment by seefeldb** [comment]
Example code:

Let's use ReAct as example and assume that we want to pass extra data to the tools.

The original, minus the schemas for `input` and `output`:
```
board
  .input("Ask ReAct").wire("text",
    board
      .include(`react-with-slot.json`, {
        slotted: { tools: tools() as unknown as NodeValue },
      })
      .wire("text", board.output())
  );
```

Let's say we have a new `location` node that provides the user's location and we want to pass that to the tools. Without changing too much of the syntax, we could just pass a `passthrough` node with the graph that collects the data.

```
board
  .input("Ask ReAct").wire("text",
    board
      .include(`react-with-slot.json`, {
        slotted: { tools: [tools() as unknown as NodeValue, board.passthrough().wire("<-location", board.location()] },
      })
      .wire("text", board.output())
  );
```

Maybe we want to remove `slotted` and make graphs just regular inputs, like so:
```
board
  .input("Ask ReAct").wire("text",
    board
      .include(`react-with-slot.json`, {
        tools: board.subboard(tools()).wire("<-location", board.location())
      })
      .wire("text", board.output())
  );
```

inside it looks like (instead of `slot`):
```... board.input().wire("tools->board", board.include(). ...```
(which is also how dynamic includes could work)

and fwiw, dynamic slots would be something like
```... board.include(`...`).wire("<-tools", some_source_of_boards) ... ```


now i want to see whether we can make `map()` and `filter()` nodes that take graphs as parameter and make inlined subgraphs easy to read. let's get as close as possible to the JS notation of `list.map((x) => { return x + a })`.

```
  board.map(board.subboard((board, input, output) => {
    const llm = board.addKit(Starter);
    input.wire(
      "text",
      llm.generateText().wire(
        "result",
        output));
  }))
  .wire("list<-", ...)
  .wire("list->", ...)
```

the nastiest bit here is that we have to add the kits again and it'll be very easy to use the main graphs `llm` here. Maybe we could make that work though? The kit is attached to the initial graph, which will know about this subgraph by the time the function is called. Once a node gets wired to the "wrong" graph, it might move it there on the fly, also making sure the kit is under the hood added to the graph?

More bikeshedding: `board.subgraph` is too long. `bind` seems incorrect. Other ideas?

**Comment by dglazkov** [comment]
Yeah, this seems very relevant right now. The idea that `slot` is just an input is very interesting.

**Comment by dglazkov** [comment]
> Maybe we want to remove `slotted` and make graphs just regular inputs, like so:

This is the key trick -- passing boards over wires as properties. I'd like to figure out how to do that.

> `board.subboard` detects a function and passes it a new board with input and output nodes as convenience

Neat trick!

**Comment by seefeldb** [comment]
> > Maybe we want to remove `slotted` and make graphs just regular inputs, like so:
> 
> This is the key trick -- passing boards over wires as properties. I'd like to figure out how to do that.

... and wiring data directly to those passed boards. That's I think the key enabler for a lot of things.

**Comment by seefeldb** [comment]
In our discussion we've identified a few next steps and remaining open questions. Let's discuss quickly and file new issues for each task:

1. We should implement something like lambdas with closures in the runner, which primarily boils down to
    - Defining subgraphs, which have their own namespaces for node ids
    - Ability to load graphs that way, in which case they are otherwise disconnected and have their own set of kits
    - Ability to define them in the context of a parent board and being able to directly wire between nodes in the lambda and the parent context (like nested closures). Here they probably share the kits.
    - Ability to pass them around, on a wire or as configuration of a node.
    - Ability to call them with a `call` node, which instantiates them. This replaces today's `include` (which we might reserve for a future use that is more Worker like) and `slot` (which is replaced by `call` on a passed lambda).
    - Hence, instead of nested graph runners, execution will happen within one runner which maintains opportunities across all lambdas/closures.
    - Calling a lambda, i.e. instantiating it, creates something like a closure, and now nodes are identified by both their id and the closure that contains them (there is also a default parent closure), so that we could in principle have the same lambda instantiated simultaneously many times and opportunities/queues would be maintained for each node separately.
       - Note: As there are wires going into lambda (definitions), we might want to track these opportunities like today, but never run them until they are instantiated. That is, when a lambda is instantiated, pending opportunities are moved over (or copied in the case of constant wires).
       - `call` passes input data to the input nodes in the lambda and outputs output data from the output nodes in the lambda, just like now.
       - TBD: Multi round. Quick sketch:
          - `call` outputs for every run of any of the `output` nodes. By default downstream routing happens via the names of the wires, but maybe we also want a higher level way to select different kinds of outputs.
          - For pending inputs, a state is returned, which is just an internal reference to closure. When passing new inputs in, it is `call`ed, instead of the initial lambda. That is, it turns into an iterator that can be called. Maybe we want a special version of call that just instantiates a lambda without passing it any inputs yet and then rely on these subsequent calls to actually send inputs?
       - Lambdas and instantiated closures serialize as state as well. Handles to running closures (see above) should serialize as well.

2. Syntax to load a graph as a lamda and wire things directly to it
     - `board.lambdaFromURL(url, options)` and possibly `board.lambdaFromJSON(json, options)` load a board as a lambda.
     - TBD: Should we allow supporting wires from/to nodes from the current or parent closures by id? This could be expressed in `options` as `allowAcrossWiring` or so.
     - Before `call`ing it, allow currying of the lambda by directly wiring data to or from it. I.e. allow `board.lambdaFromURL(...).wire("<-foo", fooSource)`.

3. Syntax to turn a single node into a lambda
     - I propose `.asLambda()` on a node, for example:
       ```js
         const lambda = llm.promptTemplate({ template: "{{foo}} is {{bar}}")
           .asLambda()
           .wire("foo<-", fooSource)
          .wire("bar<-", barSource)
       ```
       creates a lambda around a `promptTemplate` node that has `foo` and `bar` wired into it from the current (parent) 
       closure.
     - Internally this creates a lambda with an input node with a `*->` to the promptTemplate node and another `*->` to an output node.
     - Note that the position of `.asLambda()` matters a tiny bit: The `.wire()` before are directly wired to the node. Any `.wire()` after `.asLambda()` are wired to the lambda itself, so passed via its `input` to the node. Differences are small, mostly probably around whether a caller of the lamda can override that input or not (an area that is under defined anyway: What happens if the node gets a value from two wires). We'll have to play with it to make recommendations.

4. Syntax to define a whole small board as lambda
     - I propose two ways to get started here:
        - ```js
            const lambda = board.lambda();
            lambda.input().wire(
               "foo->",
               kit.someNode()
                  .wire("bar->", lambda.output());
          ```
       - ```js
            const lambda = board.lambda((board, input, output) => {
               input.wire("foo->", kit.someNode()
                 .wire("bar->", output);
            });
          ```
       -  The second makes the code (a) look like a closure and (b) exposes predefined `input` and `output` nodes as convenience, as often we wire more than one wire from/to them and thus would have to assign them to a variable. This is nicely compact.
     - The big hairy issue here is that we need to find a syntax that makes it easy and not error prone to declare whether a new node belongs into the lambda or sits outside and is just connected via a wire:
       - In the example above, `kit.someNode` will be only wired to nodes that belong the to the lambda, so we can automatically determine that it itself must belong to the the lambda.
       - But what if there was an incoming wire, e.g. `kit.someNode().wire("baz<-", bazSource).wire("bar->", output)`, with `bazSource` being defined and wired up in the parent board? To the `lambda()` method that is indistinguishable from `kit.someNode` being defined outside and just wired directly from `input` and to `output`.
       - And the behavior could be very different when such a lambda is invoked several times, especially in parallel: All the inputs from all the parallel invocations would be wired to a single instance of that node if it is defined outside, and every invocation would send its values to all the outputs at once. So this distinction really matters.
       - Two approaches:
          1. Add kits on lambda and create nodes from these local kits. E.g. in the second example, we could do something like:
              ```js
                const lambda = board.lambda((board, input, output) => {
                   const kit = board.addKit(OtherKit);
                   input.wire("foo->", kit.someNode()
                     .wire("baz<-", bazSource)
                     .wire("bar->", output);
                });
              ```
              or for convenience, if `kit` is already defined outside:
              ```js
                const lambda = board.lambda(({board, input, output, kit}) => {
                   input.wire("foo->", kit.someNode()
                     .wire("baz<-", bazSource)
                     .wire("bar->", output);
                }, { kits: { kit }});
              ```
              i.e. using the helper to create a copy of the kit
          2. Distinguish wires that cross the boundary:
              ```js
                const lambda = board.lambda((board, input, output) => {
                   input.wire("foo->", kit.someNode()
                     .wireAcross("baz<-", bazSource)
                     .wire("bar->", output);
                });
              ```
              That is, we introduce `.wireAcross` which behaves like `.wire` except that it denotes that the wire goes across closures.

           The problem with the first one is that it's very error prone: If a developer forgets to create the lambda-specific kit, then this looks valid and might even work correctly if the lambda is only ever called one after the other, but will suddenly fail in unexpected and hard to debug ways in the future.
           The second introduces new and unusual syntax, but is not only more compact, but allows us to detect errors automatically: Internally we'd trace wires from nodes that we know belong to the lambda (the input and output nodes) to other connected nodes and if any of those belong to another board/lambda, we can throw an error. We never cross `wireAcross` in that tracing, and any encountered node is added to the lambda.
           Hence I'd recommend the second syntax.

**Comment by dglazkov** [comment]
> * board.lambdaFromJSON(

`board.lamdaFromGraphDescriptor`?

**Comment by dglazkov** [comment]
> ```js
> wireAcross
> ```

fascinating.

What's a nice small chunk of work we can start on to make progress here? I'd love to turn this into a incremental progress where each step adds more clarity.

**Comment by dglazkov** [comment]
Ok, I have an idea. Let's build `schemish-generator` as a board that takes in a lambda of template-generation.

**Comment by seefeldb** [comment]
Great. I was going to propose an order of:
 1) lambda by loading a graph, passing and then calling it
 2) wiring data to the graph before passing
 3) syntax to define single node lambda inline
 4) syntax to define graph lambda inline
 5) acrossWire directly into graph

We can also move 2) back behind the two syntax ones.

Let's do that for schemish-generator?

Also, we could keep nested calls of runners for the first few, but since that comes with quite fundamental differences, I'd want to implement the closure/namespace as first as task 0)

**Comment by seefeldb** [comment]
We have enough for initial support now. #150 documents what works and is where notes for further requests should go.
Also tracking #151 and #152 for possible shorter term ergonomic improvements.
